cdap-gateway/src/test/java/co/cask/cdap/gateway/handlers/StreamHandlerTestV3.java

```diff
@@ -17,17 +17,163 @@
+
+    List<String> eventsFetchedFromStream2 = fetchEvents(streamId2);
+    Assert.assertEquals(eventsSentToStream2, eventsFetchedFromStream2);
+  }
+
+  @Test
+  public void testNamespacedMetrics() throws Exception {
+    // Create two streams with the same name, in different namespaces.
+    String streamName = "testNamespacedMetrics";
+    Id.Stream streamId1 = Id.Stream.from("namespace1", streamName);
+    Id.Stream streamId2 = Id.Stream.from("namespace2", streamName);
+
+    createStream(streamId1);
+    createStream(streamId2);
+
+    // Enqueue 10 entries to the stream in the first namespace
+    for (int i = 0; i < 10; ++i) {
+      sendEvent(streamId1, Integer.toString(i));
+    }
+
+    // Enqueue only 2 entries to the stream in the second namespace
+    for (int i = 0; i < 2; ++i) {
+      sendEvent(streamId2, Integer.toString(i));
+    }
+
+    // Check metrics to verify that the metric for events processed is specific to each stream
+    checkEventsProcessed(streamId1, 10, 10);
+    checkEventsProcessed(streamId2, 2, 10);
+  }
+
+
+  private void createStream(Id.Stream streamId) throws Exception {
+    URL url = createURL(streamId.getNamespaceId(), "streams/" + streamId.getName());
+    HttpRequest request = HttpRequest.put(url).build();
+    HttpResponse response = HttpRequests.execute(request);
+    Assert.assertEquals(200, response.getResponseCode());
+  }
+
+  private void sendEvent(Id.Stream streamId, String body) throws Exception {
+    URL url = createURL(streamId.getNamespaceId(), "streams/" + streamId.getName());
+    HttpRequest request = HttpRequest.post(url).withBody(body).build();
+    HttpResponse response = HttpRequests.execute(request);
+    Assert.assertEquals(200, response.getResponseCode());
+  }
+
+  private List<String> fetchEvents(Id.Stream streamId) throws Exception {
+    URL url = createURL(streamId.getNamespaceId(), "streams/" + streamId.getName() + "/events");
+    HttpRequest request = HttpRequest.get(url).build();
+    HttpResponse response = HttpRequests.execute(request);
+    Assert.assertEquals(200, response.getResponseCode());
+
+
+    List<String> events = Lists.newArrayList();
+    JsonArray jsonArray = new JsonParser().parse(response.getResponseBodyAsString()).getAsJsonArray();
+    for (JsonElement jsonElement : jsonArray) {
+      events.add(jsonElement.getAsJsonObject().get("body").getAsString());
+    }
+    return events;
+  }
+  
+  private void checkEventsProcessed(Id.Stream streamId, int expectedCount, int retries) throws Exception {
+    for (int i = 0; i < retries; i++) {
+      int numProcessed = getNumProcessed(streamId);
+      if (numProcessed == expectedCount) {
+        return;
+      }
+      TimeUnit.SECONDS.sleep(1);
+    }
+    Assert.assertEquals(expectedCount, getNumProcessed(streamId));
+  }
+
+
+  private int getNumProcessed(Id.Stream streamId) throws Exception {
+    String path = String.format("/v3/metrics/query?metric=system.collect.events&context=ns.%s.str.%s",
```

Comment:
pull latest develop, after https://github.com/caskdata/cdap/pull/1365 
we use user-friendly tag name now, so it would be like , namespace.%s.stream.%s